src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ResettableValueTaskSource.cs

```diff
@@ -130,84 +133,102 @@ public bool TryGetValueTask(out ValueTask valueTask, object? keepAlive = null, C
         }
     }
 
-    public Task GetFinalTask() => _finalTaskSource.Task;
+    /// <summary>
+    /// Gets a <see cref="Task"/> that will transition to a completed state with the last transition of this source, i.e. into <see cref="State.Completed"/>.
+    /// </summary>
+    /// <returns>The <see cref="Task"/> that will transition to a completed state with the last transition of this source.</returns>
+    public Task GetFinalTask(object? keepAlive)
+    {
+        lock (this)
+        {
+            return _finalTaskSource.GetTask(keepAlive);
+        }
+    }
 
     private bool TryComplete(Exception? exception, bool final)
     {
+        // Dispose the cancellation registration before completing the task, so that it cannot run after the awaiting method returned.
```

Comment:
Solves #82704.
As the registration was disposed after completing the tasks, they continuation was allowed to run, which reset this value task source while cancellation registration was still alive and could fire before the disposal.